zebedee-cms/src/main/java/com/github/onsdigital/zebedee/util/versioning/VersionsServiceImpl.java

```diff
@@ -0,0 +1,177 @@
+        Integer version = -1;
+        Optional<String> vOpt = getVersionNameFromURI(uri);
+
+        if (vOpt.isPresent()) {
+            version = Integer.valueOf(vOpt.get().replaceFirst("v", ""));
+        }
+
+        return version;
+    }
+
+    @Override
+    public Comparator<Version> versionComparator() {
+        return (Version v1, Version v2) -> {
+            Integer foo = getVersionNumberFromURI(v1.getUri().toString());
+            Integer bar = getVersionNumberFromURI(v2.getUri().toString());
+            return foo.compareTo(bar);
+        };
+    }
+
+    @Override
+    public boolean isVersionDir(File file) {
+        return file != null && file.isDirectory() && VALID_VERSION_DIR_PATTERN.matcher(file.getName()).matches();
+    }
+
+
+    @Override
+    public void verifyCollectionDatasets(ZebedeeReader cmsReader, Collection collection,
+                                         CollectionReader reader, Session session)
+            throws ZebedeeException, IOException, VersionNotFoundException {
+
+        List<Dataset> datasets = getDatasetsInCollection(reader);
+
+        if (datasets != null && !datasets.isEmpty()) {
+            verifyDatasetVersions(cmsReader, collection, session, datasets);
+
+            info().collectionID(collection)
+                    .user(session)
+                    .data("datasets",
+                            datasets.stream()
+                                    .map(ds -> ds.getDescription().getTitle())
+                                    .collect(Collectors.toList()))
+                    .log("collection legacy datasets verified successfully");
+        }
+    }
+
+    List<Dataset> getDatasetsInCollection(CollectionReader reader) throws ZebedeeException, IOException {
+        List<Dataset> datasets = new ArrayList<>();
+
+        for (String uri : reader.getReviewed().listUris()) {
+
+            if (isDataJson(uri) && !isVersionedURI(uri)) {
+
+                Path parent = Paths.get(uri).getParent();
+
+                Page page = reader.getReviewed().getContent(parent.toString());
+                if (PageType.dataset.equals(page.getType())) {
+                    datasets.add((Dataset) page);
+                }
+            }
+        }
+
+        return datasets;
+    }
+
+    void verifyDatasetVersions(ZebedeeReader cmsReader, Collection collection, Session session,
+                               List<Dataset> datasets) throws ZebedeeException, IOException,
+            VersionNotFoundException {
+        for (Dataset ds : datasets) {
+            info().collectionID(collection)
+                    .data("dataset", ds.getDescription().getTitle())
+                    .log("verifying collection dataset versions");
+
+            for (Version version : ds.getVersions()) {
+                String uri = version.getUri().toString();
+
+                try {
+                    cmsReader.getCollectionContent(collection.getId(), session.getId(), uri);
```

Comment:
Consider checking also if the excel file exists for the dataset if its found in the collection